File: haskell/curve/Interpolation.hs
module Interpolation (
    lower_bound
  , upper_bound
  , equal_range
  , bound
  , linear_interpolation
  , loglinear_interpolation)
where

-- | lower_bound
-- | 
-- | Finds the first position in xs where x could be inserted
-- | without violating the ordering.

lower_bound :: (Ord a) => a -> [a] -> Int
lower_bound_ x xs first count
  | count == 0      = first
  | (xs !! mid < x) = lower_bound_ x xs (mid+1) (count-half-1)
  | otherwise       = lower_bound_ x xs first half 
  where half = count `div` 2;  mid = first + half

lower_bound x xs  = lower_bound_ x xs 0 (length xs)

-- | upper_bound
-- | 
-- | Finds the last position in xs where x could be inserted
-- | without violating the ordering.

upper_bound :: (Ord a) => a -> [a] -> Int
upper_bound_ x xs first count
   | count == 0         = first
   |  x < (xs !! mid)   = upper_bound_ x xs first half
   | otherwise          = upper_bound_ x xs (mid + 1) (count-half - 1)
  where half = count `div` 2;  mid = first + half

upper_bound x xs  = upper_bound_ x xs 0 (length xs)

-- | equal_range
-- | 
-- | Finds the largest subrange in which x could be inserted in any
-- | place without violating the ordering.

equal_range :: (Ord a) => a -> [a] -> (Int, Int)
equal_range x xs = (lower_bound x xs, upper_bound x xs)

-- | bound
-- | 
-- | Error if x is outside of the domain otherwise find indicies i, j
-- | such that xs!!i <= x <= xs!!(i+1).

bound x xs
  | left == count   = error ("bound :" ++ show x ++ " left of domain")
  | right == 0      = error ("bound :" ++ show x ++ " right of domain")
  | right == count  = (left, right-1)
  | left == right   = (left-1, right)
  | otherwise       =   (left, right)
  where count = length(xs); (left, right) = equal_range x xs

-- | linear_interpolation
-- | 
-- | Error if x is outside of the domain otherwise linearly interpolate.

linear_interpolation :: (Floating a, Ord a) => a -> [a] -> [a] -> a
linear_interpolation x xs ys = 
    let 
      (i', i)   = bound x xs
      (xi', xi) = (xs!!i', xs!!i)
      (yi', yi) = (ys!!i', ys!!i)
      r         = (x - xi)/(xi' - xi)
    in  r*(yi' - yi) + yi

-- | loglinear_interpolation
-- | 
-- | Error if x is outside of the domain otherwise log-linearly interpolate.

loglinear_interpolation :: (Floating a, Ord a) => a -> [a] -> [a] -> a
loglinear_interpolation x xs ys = 
    let 
      (i', i)   = bound x xs
      (xi', xi) = (xs!!i', xs!!i)
      (yi', yi) = (ys!!i', ys!!i)
      r         = (x - xi)/(xi' - xi)
      lyi       = log yi
      lyi'      = log yi'
    in exp(r*(lyi' - lyi) + lyi)

File: ocaml/curve/interpolation.ml
module Interpolation =
  struct

    let lower_bound : float -> float list -> int = 
      fun x xs ->
	let rec loop = fun first count -> 
	  if count = 0 then
	    first
	  else
	    let half = count/2
	    in 
  	      let mid = first+half
	      in 
	        if (List.nth xs mid) < x then
	          loop (mid+1) (count-half-1)
	        else
	          loop first half
	in
	  loop 0 (List.length xs)
    ;;
 
    let upper_bound : float -> float list -> int = 
     fun x xs ->
       let rec loop = fun first count ->
	 if count = 0 then first
	 else
	   let half=count/2
	   in
  	     let mid=first+half
	     in
	     if x < (List.nth xs mid) then
	       loop first half
	     else
	       loop (mid+1) (count-half-1)
       in
        loop 0 (List.length xs)
    ;;

    let equal_range : float -> float list -> (int*int) = 
      fun x xs ->
	let a = lower_bound x xs
	and b = upper_bound x xs
	in 
          (a, b)
    ;;

  end
;;

open Interpolation ;;

let p = Interpolation.equal_range 3.0 [1.0; 2.0; 4.0] in
  Printf.printf "lower_bound 3.0 [1.0; 2.0; 4.0] = %d\n" (Interpolation.lower_bound 3.0 [1.0; 2.0; 4.0]) ;
  Printf.printf "upper_bound 3.0 [1.0; 2.0; 4.0] = %d\n" (Interpolation.upper_bound 3.0 [1.0; 2.0; 4.0]) ;
  Printf.printf "equal_range 3.0 [1.0; 2.0; 4.0] = (%d, %d)\n" (fst p) (snd p)

